Automattic/woocommerce-payments
# Fix changing payment to saved payment for subscriptions order

Fixes #1185

#### Changes proposed in this Pull Request

* Override `WC_Payment_Gateway_WCPay::process_payment()` in `WC_Payment_Gateway_WCPay_Subscriptions_Compat::process_payment()` to handle specifically payment method change for subscription order, because:
  * [On caught exception, it updates the order's status to failed](https://github.com/Automattic/woocommerce-payments/blob/trunk/includes/class-wc-payment-gateway-wcpay.php#L470).
  * Taking out [this return statement](https://github.com/Automattic/woocommerce-payments/blob/62516cdcee37993cae25999c6256d5c574e53124/includes/class-wc-payment-gateway-wcpay.php#L571-L574) per [suggested here](https://github.com/Automattic/woocommerce-payments/pull/1148#issuecomment-750410406) is not enough because [stock will be reduced and cart will be emptied on success](https://github.com/Automattic/woocommerce-payments/blob/62516cdcee37993cae25999c6256d5c574e53124/includes/class-wc-payment-gateway-wcpay.php#L711-L714).
  * `WC_Payment_Gateway_WCPay` shouldn't know the nuts and bolts of subscription, so doing something like "if changing payment method for subscription, do not reduce stock", for example, IMO breaks the abstraction.

#### Testing instructions

Shamelessly copied from [1185](https://github.com/Automattic/woocommerce-payments/issues/1185).
1. Purchase subscription product.
2. Go to My account > Add one more payment method.
3. Go to My account > Subscription > Purchased subscription.
4. Try to change payment method to saved on step 2.
5. Success message is displayed. With `trunk`, payment method is not changed. With this branch, payment method should be changed.



-------------------
